### PR TITLE
File IO: File open/save dialog utility functions

### PR DESCRIPTION
**Pull request by Scepheo**
Repeat of [PR 61](https://github.com/Hourglass-Resurrection/Hourglass-Resurrection/pull/61), as it turns out these changes still apply after the switch to C++17.

### CODE REVIEW

File: source/application/ramwatch.cpp [hunk 1]
```diff
@@ -444,64 +447,24 @@ void OpenRWRecentFile(int memwRFileNumber)
...
+    });
 
-	if(GetOpenFileName(&ofn)) return 1;
-
-	return 0;
+    file_name.copy(Dest, file_name.length());
+    return file_name.empty() ? 0 : 1;
 }
 
 int Change_File_S(char *Dest, const char *Dir, const char *Titre, const char *Filter, const char *Ext, HWND hwnd)
```
Comment (Warepire) [inline source/application/ramwatch.cpp]: Same here.

File: source/application/ramwatch.cpp [hunk 2]
```diff
@@ -444,64 +447,24 @@ void OpenRWRecentFile(int memwRFileNumber)
 
 int Change_File_L(char *Dest, const char *Dir, const char *Titre, const char *Filter, const char *Ext, HWND hwnd)
```
Comment (Warepire) [inline source/application/ramwatch.cpp]: Many of these params are never used now, perhaps reduce their signatures?

File: source/application/utils/file/FileDialog.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            const char* filter = file_type_map[file_type];
+            result.append(filter);
+        }
+
+        std::replace(result.begin(), result.end(), '|', '\0');
+        return result;
+    }
+}
+
+std::string Utils::File::GetFileNameOpen(
```
Comment (Warepire) [inline source/application/utils/file/FileDialog.cpp]: Oh the code duplication between Open and Save. Can you consolidate this? Perhaps with an `enum class` to pick Open or Save?

**Review by Warepire — CHANGES_REQUESTED** [review]

File: source/application/utils/file/FileDialog.cpp [hunk 4]
```diff
@@ -38,46 +38,45 @@ namespace
...
         return result;
     }
+
+    typedef BOOL(WINAPI FileNameGetter)(LPOPENFILENAME open_file_name);
+
+    std::string GetFileName(
+        const std::string& location,
+        const std::vector<Utils::File::FileFilter>& file_types,
+        const DWORD& flags,
+        const FileNameGetter get)
```
Comment (Warepire) [inline source/application/utils/file/FileDialog.cpp]: Rename `get` to something more clear, like `file_name_getter`.

File: source/application/utils/file/FileDialog.cpp [hunk 5]
```diff
@@ -38,46 +38,45 @@ namespace
         std::replace(result.begin(), result.end(), '|', '\0');
         return result;
     }
+
+    typedef BOOL(WINAPI FileNameGetter)(LPOPENFILENAME open_file_name);
```
Comment (Warepire) [inline source/application/utils/file/FileDialog.cpp]: Use `using`.

**Review by Warepire — CHANGES_REQUESTED** [review]

**Comment by Warepire** [comment]
Please validate that it still works, and once it does, please squash the commits.

Code looks good to me.

**Comment by Scepheo** [comment]
It works on my machine.

Verified, it appears to still work as before. I squashed the commits, feel free to merge the PR if you see fit.

**Review by Warepire — APPROVED** [review]

**Pull request merged**